File: 0062-unique-paths/0062-unique-paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        
        @lru_cache(maxsize = None)
        def dp(i: int, j: int) -> int:
            if i + j == 0:
                return 1
            
            ways = 0
            if i > 0:
                ways += dp(i - 1, j)
            
            if j > 0:
                ways += dp(i, j - 1)
            
            return ways
        
        return dp(m - 1, n - 1)
        

File: 0062-unique-paths/0062-unique-paths.cpp
class Solution {
private:
    int dp(int row, int col, vector<vector<int>>& memo){
        if (row + col == 0) return 1;
        
        if (memo[row][col] != -1) return memo[row][col];
        
        int ways = 0;
        if (row > 0) ways += dp(row - 1, col, memo);
        if (col > 0) ways += dp(row, col - 1, memo);
        
        return memo[row][col] = ways;
    }
    
public:
    int uniquePaths(int m, int n) {
        vector<vector<int>> memo(m, vector<int>(n, -1));
        
        return dp(m - 1, n - 1, memo);
    }
};